File: src/transfer.ts
import { fetchLinearIssues, LinearIssue } from "./linear";
import { createOctokit, createGitHubIssue, GitHubIssue } from "./github";

export interface TransferOptions {
  linearApiKey: string;
  githubToken: string;
  githubOwner: string;
  githubRepo: string;
  linearTeamId?: string;
  delayMs?: number;
}

export interface TransferResult {
  linearIssue: LinearIssue;
  githubIssue: GitHubIssue;
}

export interface TransferSummary {
  total: number;
  transferred: number;
  failed: number;
  results: TransferResult[];
  errors: Array<{ issue: LinearIssue; error: string }>;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function transferIssues(
  options: TransferOptions,
  onProgress?: (
    current: number,
    total: number,
    issue: LinearIssue
  ) => void
): Promise<TransferSummary> {
  const {
    linearApiKey,
    githubToken,
    githubOwner,
    githubRepo,
    linearTeamId,
    delayMs = 1000,
  } = options;

  const linearIssues = await fetchLinearIssues(linearApiKey, linearTeamId);
  const octokit = createOctokit(githubToken);

  const summary: TransferSummary = {
    total: linearIssues.length,
    transferred: 0,
    failed: 0,
    results: [],
    errors: [],
  };

  for (let i = 0; i < linearIssues.length; i++) {
    const issue = linearIssues[i];

    if (onProgress) {
      onProgress(i + 1, linearIssues.length, issue);
    }

    try {
      const githubIssue = await createGitHubIssue(
        octokit,
        githubOwner,
        githubRepo,
        issue
      );
      summary.results.push({ linearIssue: issue, githubIssue });
      summary.transferred++;
    } catch (err) {
      const message = err instanceof Error ? err.message : String(err);
      summary.errors.push({ issue, error: message });
      summary.failed++;
    }

    if (i < linearIssues.length - 1 && delayMs > 0) {
      await sleep(delayMs);
    }
  }

  return summary;
}

File: cli-go/transfer.go
package main

import (
	"context"
	"fmt"
	"time"
)

type TransferOptions struct {
	LinearAPIKey string
	GitHubToken  string
	GitHubOwner  string
	GitHubRepo   string
	LinearTeamID string
	DelayMs      int
}


type TransferResult struct {
	LinearIssue LinearIssue
	GitHubIssue GitHubIssue
}


type TransferError struct {
	Issue LinearIssue
	Err   string
}

type TransferSummary struct {
	Total       int
	Transferred int
	Failed      int
	Results     []TransferResult
	Errors      []TransferError
}

func transferIssues(opts TransferOptions, onProgress func(current, total int, issue LinearIssue)) (*TransferSummary, error) {
	fmt.Println("Fetching Linear issues...")
	issues, err := fetchLinearIssues(opts.LinearAPIKey, opts.LinearTeamID)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch Linear issues: %w", err)
	}
	fmt.Printf("Found %d issues. Starting transfer...\n", len(issues))

	ctx := context.Background()
	summary := &TransferSummary{Total: len(issues)}

	for i, issue := range issues {
		if onProgress != nil {
			onProgress(i+1, len(issues), issue)
		}

		ghIssue, err := createGitHubIssue(ctx, opts.GitHubToken, opts.GitHubOwner, opts.GitHubRepo, issue)
		if err != nil {
			fmt.Printf("  Failed: %v\n", err)
			summary.Failed++
			summary.Errors = append(summary.Errors, TransferError{Issue: issue, Err: err.Error()})
		} else {
			fmt.Printf("  Created: %s\n", ghIssue.URL)
			summary.Transferred++
			summary.Results = append(summary.Results, TransferResult{LinearIssue: issue, GitHubIssue: *ghIssue})
		}

		if i < len(issues)-1 && opts.DelayMs > 0 {
			time.Sleep(time.Duration(opts.DelayMs) * time.Millisecond)
		}
	}

	return summary, nil
}
